File: fractal/src/main/java/net/fwitz/math/fractal/escape/EscapeTimeResult.java
package net.fwitz.math.fractal.escape;

import net.fwitz.math.binary.complex.Complex;

import java.util.OptionalDouble;

import static java.util.Objects.requireNonNull;

public class EscapeTimeResult {
    private final boolean escaped;
    private final int iters;
    private final Complex z;
    private final OptionalDouble smoothing;

    private EscapeTimeResult(boolean escaped, int iters, Complex z, OptionalDouble smoothing) {
        this.escaped = escaped;
        this.iters = iters;
        this.z = requireNonNull(z, "z");
        this.smoothing = requireNonNull(smoothing, "smoothing");
    }

    public static EscapeTimeResult contained(Complex z, OptionalDouble smoothing) {
        return new EscapeTimeResult(false, 0, z, smoothing);
    }

    public static EscapeTimeResult escaped(int iters, Complex z, OptionalDouble smoothing) {
        return new EscapeTimeResult(true, iters, z, smoothing);
    }

    public int iters() {
        return iters;
    }

    public Complex z() {
        return z;
    }

    public boolean escaped() {
        return escaped;
    }

    public boolean contained() {
        return !escaped;
    }

    public OptionalDouble smoothing() {
        return smoothing;
    }

    @Override
    public String toString() {
        return "EscapeFunction.Result[escaped=" + escaped +
                "; iters=" + iters +
                "; z=" + z +
                "; smoothing=" + smoothing +
                ']';
    }
}

File: fractal/src/main/kotlin/net/fwitz/math/fractal/escape/EscapeTimeResult.kt
package net.fwitz.math.fractal.escape

import net.fwitz.math.binary.BinaryNumber

class EscapeTimeResult<T : BinaryNumber<T>> private constructor(
    val escaped: Boolean,
    val iters: Int,
    val z: T,
    val smoothing: Double?
) {
    companion object {
        fun <T : BinaryNumber<T>> contained(z: T, smoothing: Double?): EscapeTimeResult<T> {
            return EscapeTimeResult(false, 0, z, smoothing)
        }

        fun <T : BinaryNumber<T>> escaped(iters: Int, z: T, smoothing: Double?): EscapeTimeResult<T> {
            return EscapeTimeResult(true, iters, z, smoothing)
        }
    }

    val contained get() = !escaped

    override fun toString() = "EscapeFunction.Result[escaped=$escaped; iters=$iters; z=$z; smoothing=$smoothing]"
}